Step 2 코어데이터 DB 구현 - 이니, 글렌

iluxsm opened the pull request:
<kbd>@myssun0325 ❤️🧡💛💚💙남💜</kbd>

# 주요 구현 사항
- 코어데이터 Entity 생성했습니다.
- CRUD 구현했습니다.
    - CREATE -> 기본 더미 데이터(title: 새로운 메모, body: 추가 텍스트 없음) DB 저장 -> cell 추가 -> cell에 대응하는 메모 타입이 `showDetailViewController`(ContentViewController) 를 통해 표시
- `NSAttributedString`을 활용하여 제목과 본문에 속성을 적용시켰고, 둘을 첫 번째 `\n` 으로 구분했습니다.
- 더 보기 버튼을 탭 했을 때, 공유하기 및 삭제하기 기능 구현했습니다.
- ContentViewController의 NavigationViewController가 새로운 cell을 클릭할 때 마다 스택에 쌓이지 않게 `popToRootViewController` -> pushViewController 순으로 진행되도록 구현했습니다.
- ListViewController에서 ContentViewController로 넘어갈 때는 showDetailViewController로, 반대의 경우에는 델리게이션 패턴을 활용해서 ListViewController에서 CRUD를 전부 구현했습니다.

# 고민할 점 및 고쳐야할 점
> 메이슨, 도움!! 🧠
- Memo를 작성하러 진입했다가 수정하지 않고 리스트로 돌아올 때, 저장된 더미 데이터 및 생성된 셀을 삭제처리 되도록 해야할 것 같습니다.
- CRUD의 UPDATE 구현부에서 `if context.hasChanges {}` 컨텐트 내용이 변경되지 않아도 동작해서 불필요한 저장 과정이 진행됩니다.
- CRUD만 따로 `MemoDataManager` 타입으로 분리하면 좋을 것 같습니다.
- 문자열 리터럴을 관리하는 타입 만드는 것에 대해 고민 중입니다.
- ContentViewController의 `didTapOption` 메소드의 기능을 세분화하여 분리하는 것이 좋을 것 같습니다.
- 더 보기 버튼을 누르면 오토레이아웃 디버그 관련 로그가 뜨는데, 수정해야할 것 같습니다.
- 더보기 팝업창이 뜬 상태에서 화면이 portrait mode 와 landscape mode 간 전환 시 팝업 위치가 가운데를 못 찾는 것 같습니다.

Commit:
[#10] chore: 기존 메모 모델 삭제 및 코어 데이터 세팅

Commit:
[#10] chore: 기존 메모 모델 관련 메서드 및 오류 구문 주석 처리

innieminnie committed:
[#12] feat: AlertController 기본 구현

Commit:
Merge pull request #13 from innieminnie/step2

[#12] feat: AlertController 기본 구현

innieminnie committed:
[#12] fix: 더보기 버튼 이미지 수정

innieminnie committed:
[#12] feat: activityView 관련 기능 추가

innieminnie committed:
[#12] refactor: activityView의 presenting 방식 수정

innieminnie committed:
[#12] feat: UIActivityViewController의 activityItems 추가

Commit:
[#9] refactor: 불피요한 오버라이딩 제거

Commit:
Merge pull request #14 from innieminnie/step2-glenn

Step2 glenn

Commit:
[#11] Create, Read 구현

Commit:
[#11] Create, Read 구현

Commit:
Merge pull request #15 from innieminnie/step2-glenn

Step2 glenn

innieminnie committed:
Merge branch 'step2-innie' into step2

innieminnie committed:
[#11] feat: 메모 삭제 기능 (CoreData) 구현

innieminnie committed:
[#11] chores: 코드 위치 일부 수정

innieminnie committed:
[#11] refactor: delete기능 Delegation 패턴 활용 방식으로 변경

innieminnie committed:
[#11] fix: ContentViewController의 temporaryMemo 초기화 오류 해결

Commit:
[#11] Update 구현 (미완성)

Commit:
[#11] feat: Update 구현

Commit:
Merge pull request #16 from innieminnie/step2-glenn

Step2 glenn

innieminnie committed:
[#17] refactor: NavigationController popToRootViewController 활용

innieminnie committed:
[#18] chores: 코드 정리

innieminnie committed:
[#18] feat: 새로운 메모 작성 취소 시 메모 삭제되는 기능 구현

innieminnie committed:
[#18] fix: 업데이트 내용 없을 시 추가적인 save 동작 막도록 수정

innieminnie commented:
위에 작성한 고민사항 중, 리팩토링 작업을 거쳐 구현한 부분에 대해 추가적으로 push합니다.

- Memo를 작성하러 진입했다가 수정하지 않고 리스트로 돌아올 때, 저장된 더미 데이터 및 생성된 셀을 삭제처리 되도록 해야할 것 같습니다.
  -> ListViewController로 전달된 memo의 내용을 updateMemo()에서 확인한 후 작성된 내용이 없을 시 deleteItem을 수행, 작성된 내용이 있을 시엔 updateItem을 수행하도록 했습니다.

- CRUD의 UPDATE 구현부에서 if context.hasChanges {} 컨텐트 내용이 변경되지 않아도 동작해서 불필요한 저장 과정이 진행됩니다.
  ->  ListViewController의 updateItem에서 조건문으로 작성한 context.hasChanges는 실질적으로 내용이 바뀌냐에 따라 bool값이 바뀌는게 아니라 NSManagedObject에 접근함과 동시에 true로 설정된다는 것을 확인했습니다. 따라서 해당 조건문을 if !(memo.changedValues().isEmpty) 로 변경하여 changedValues의 element의 수를 확인하는 작업을 추가했고, memo의 lastModified는 해당 조건문이 true일 때 현재시간으로 update하여 CoreData에 반영되도록, false일 시에는 title, body, lastModified가 기존값으로 유지된 채 조건문 내의 save() 과정을 거치지 않도록 구현했습니다.

innieminnie committed:
[#18] fix: 공유하기 팝업창 디스플레이 모드 전환 시 위치 찾기 수정

innieminnie commented:
- ContentViewController의 didTapOption 메소드의 기능을 세분화하여 분리하는 것이 좋을 것 같습니다.
- 더보기 팝업창이 뜬 상태에서 화면이 portrait mode 와 landscape mode 간 전환 시 팝업 위치가 가운데를 못 찾는 것 같습니다.

didTapOption 내의 Action별로 프로퍼티 생성, viewWillTransition 메소드를 활용을 통해 위의 두 고민점 또한 수정하여 구현해보았습니다. 참고 부탁드립니다 :)

Commit:
[#18] refactor: StringCollection 생성 및 문자열 리터럴 교체

Commit:
Merge branch 'step2' into step2-glenn

Commit:
Merge pull request #20 from innieminnie/step2-glenn

[#18] refactor: StringCollection 생성 및 문자열 리터럴 교체

iluxsm commented:
# 추가 사항
- 문자열 리터럴을 사용하던 것을 StringCollection을 따로 만들어서 관리하도록 변경했습니다. 그런데, 기준을 잘 잡은 것인지 잘 모르겠네요. 개인적으로 문자열 리터럴이 변경될 수 있다고 생각하는 것들을 카테고리로 모아서 여러 객체로 만들어 놨는데, 현명한 선택인지 잘 모르겠습니다.
- nil-coalescing 연산자 또는 Default로 빈 문자열("") 이 들어가는 것들을 String()으로 변경해줬습니다. 사실 이 것에 대해서 [검색](https://stackoverflow.com/questions/31197008/swift-string-via-string-literal-vs-initializer) 해봤을 때, 내부적인 동작으로도 두 가지 방법이 동일한 것이다라는 것을 알게됐습니다. 하지만, 문자열 리터럴을 따로 관리해주는 입장에서 통일성(?)을 위해 해줬습니다. 이 부분에 대해서 메이슨의 생각이 궁금하네요.

innieminnie committed:
[#18] fix: revert를 통한 버그 수정

myssun0325 commented on CloudNotes/CloudNotes/Cells/MemoListCell.swift:
```diff
@@ -48,9 +48,9 @@ class MemoListCell: UITableViewCell {
-        titleLabel.text = ""
-        dateLabel.text = ""
-        predescriptionLabel.text = ""
+        titleLabel.text = String()
```
빈 문자열을 매번생성하는 것과 emptyString같이 변수로 관리하는것과 어떤차이가 있을까요?

myssun0325 commented on CloudNotes/CloudNotes/Controllers/ContentViewController.swift:
```diff
@@ -0,0 +1,244 @@
+    
+    private var popoverController: UIPopoverPresentationController?
+    private var activityViewController: UIActivityViewController?
+    private lazy var alertController: UIAlertController = {
```
개행의 기준도 있어야하지 않을까요?

myssun0325 commented on CloudNotes/CloudNotes/Controllers/ContentViewController.swift:
```diff
@@ -0,0 +1,244 @@
+    private let headLinefont = UIFont.boldSystemFont(ofSize: 24)
+    private let bodyLinefont = UIFont.systemFont(ofSize: 15)
+    private var currentMemo: Memo?
+    var delegate: MemoDelegate?
```
애플 공식문서에 UITableView의 delegate 속성이 어떻게 되어있는지 참고해볼까요~?
델리게이트의 경우 순환참조 때문에 `weak`로 주로 약한참조를 사용하곤합니다.
weak를 사용하기 위해선 Delegate프로토콜이 어떻게 바껴야할까용?

iluxsm replied:
AnyObject 프로토콜을 준수해야합니당

myssun0325 commented on CloudNotes/CloudNotes/Controllers/ContentViewController.swift:
```diff
@@ -0,0 +1,244 @@
+    }()
+    
+    private lazy var shareAction: UIAlertAction = {
+        let shareAction = UIAlertAction(title: "Share...", style: .default, handler: { (alert: UIAlertAction!) -> Void in
```
"Share..."와 같은 문자열도 의미있는 변수나 상수로 나타내는건 어떨까요?

myssun0325 commented on CloudNotes/CloudNotes/Controllers/ContentViewController.swift:
```diff
@@ -0,0 +1,244 @@
+            
+            self.popoverController = popoverController
+            popoverController.sourceView = self.view
+            popoverController.sourceRect = CGRect(x: self.view.bounds.midX, y: self.view.bounds.midY, width: 0, height: 0)
```
실제 view의 정확한 view의 bounds가 잡히는 시점은 언제일까요?~ 
bounds나 frame을 다룰 땐 시점을 고려하는걸 잊지마세용 :)

iluxsm replied:
오토 레이아웃을 설정하고, 실행될 때 뷰의 `bounds`와 `frame`이 설정된다고 합니다.

iluxsm replied:
정확하게는 잘 모르겠지만, `frame` 부모 뷰의 좌표계를 사용하여 위치를 잡고, `bounds`는 부모 뷰의 좌표와 관계없이 자신만의 독자적인 좌표계를 사용하여 위치를 잡는 것이라고 알고 있습니다. 그 말로 미루어봤을 때, `frame`은 부모 뷰가 이니셜라이즈 된 시점에서는 해당 뷰의 이니셜라이즈와 관계없이 사용할 수 있을 것 같고, `bounds`는 스스로 이니셜라이즈가 된 이후에 사용할 수 있을 것 같습니다.

myssun0325 commented on CloudNotes/CloudNotes/Controllers/ContentViewController.swift:
```diff
@@ -0,0 +1,244 @@
+        contentView.isScrollEnabled = false
+        contentView.isEditable = false
+        contentView.autocorrectionType = .no
+        contentView.isUserInteractionEnabled = true
```
이미 같은 값으로 설정되어있는 건 불필요하게 재설정하지 않아도 될듯해용 :)

myssun0325 commented on CloudNotes/CloudNotes/Controllers/ContentViewController.swift:
```diff
@@ -0,0 +1,244 @@
+        }
+    }
+    
+    override func viewWillTransition(to size: CGSize, with coordinator: UIViewControllerTransitionCoordinator) {
```
요 메서드는 super호출이 빠져있네용 @@
`viewWillTransition`은 언제 호출되는 메서드일까요?

iluxsm replied:
정확하게는 View의 사이즈가 변경될 때이고, 그 중 하나로 Portrait 와 Landscape 모드가 서로 전환될 때 호출되는 시점을 체크하기 위해 사용했습니다.

myssun0325 commented on CloudNotes/CloudNotes/Controllers/ContentViewController.swift:
```diff
@@ -0,0 +1,244 @@
+        let startIndex: String.Index = contentView.text.startIndex
+        guard let endIndex: String.Index = contentView.text.firstIndex(of: "\n") else { return Memo() }
+        let afterEndIndex: String.Index = contentView.text.index(after: endIndex)
+        
```
개행이 두번됐네용

myssun0325 commented on CloudNotes/CloudNotes/Controllers/ListViewController.swift:
```diff
@@ -0,0 +1,162 @@
+    }
+}
+extension ListViewController: MemoDelegate {
+    func updateMemo(memo: Memo) {
```
메모추가에 들어가서 지운다음에 뒤로가기를 누르면 요 메서드에서 죽네용

myssun0325 commented on CloudNotes/CloudNotes/Controllers/ListViewController.swift:
```diff
@@ -0,0 +1,162 @@
+}
+extension ListViewController: MemoDelegate {
+    func updateMemo(memo: Memo) {
+        guard memo.title != String(), memo.body != String() else {
```
memo.title은 옵셔널이네요. 어떤 비교가 좋을까요?
아 참고로 String의 isEmpty라는 메서드가 무엇인지 찾아볼까요?

myssun0325 commented on CloudNotes/CloudNotes/Controllers/ListViewController.swift:
```diff
@@ -0,0 +1,162 @@
+
+class ListViewController: UITableViewController {
+    private let context = (UIApplication.shared.delegate as! AppDelegate).persistentContainer.viewContext
+    private var memoList = [Memo]()
```
모델을 관리하는 객체를 따로 만들어서 역할을 분리하는건 어떨까요

myssun0325 commented on CloudNotes/CloudNotes/Controllers/ListViewController.swift:
```diff
@@ -0,0 +1,162 @@
+            return UITableViewCell()
+        }
+        
+        if memoList[indexPath.row].title == String(),
```
퀴즈! title이 옵셔널인데 어떻게 비교가 가능할까요?

iluxsm replied:
[참고](https://developer.apple.com/documentation/swift/optional)


두 연산자( `==`, `!=`) 는
1. 값과 값의 비교
2. nil 과 옵셔널끼리의 비교

를 지원한다고 합니다. 공식 문서에서 논 옵셔널 타입은 옵셔널 타입과 비교를 들어갈 때 래핑된다고 합니다.

> You can also use this operator to compare a non-optional value to an optional that wraps the same type. The non-optional value is wrapped as an optional before the comparison is made.

myssun0325 commented on CloudNotes/CloudNotes/Controllers/ListViewController.swift:
```diff
@@ -0,0 +1,162 @@
+            try context.save()
+            fetchMemo()
+        } catch(let error) {
+            #if DEBUG
```
디버그모드가 아닌경우에 에러가 아니라면 어떻게 처리해줘야할까요?

iluxsm replied:
사용자에게는 해당 작업에 실패했음을 알리는 창을 `UIAlertController` 를 통해 생성해서 보여줘야할 것 같네요!

myssun0325 commented on CloudNotes/CloudNotes/Controllers/ListViewController.swift:
```diff
@@ -0,0 +1,162 @@
+    private func fetchMemo() {
+        do {
+            memoList = try context.fetch(Memo.fetchRequest())
+            DispatchQueue.main.async {
```
클로져 안에서 self를 사용하는 경우 순환참조가 일어날 수 있습니다 어떻게 해결하면 좋을까요?

iluxsm replied:
Escaping 이기 때문에 캡쳐를 위해 `self` 를 사용해야하는데, 클로저 내부에서 순환참조를 해결하기 위해 캡쳐 리스트를 사용하면 되는 것으로 알고있습니다! 하지만, 단순히 `[weak self]` 가 아직 정확하게 무엇을 의미하는지 몰라서, 공부 해볼게요!

myssun0325 commented on CloudNotes/CloudNotes/Controllers/ListViewController.swift:
```diff
@@ -0,0 +1,162 @@
+            contentVC.didTapMemoItem(with: recentlyCreatedMemo)
+        }
+
+        (splitViewController?.viewControllers.last as? UINavigationController)?.popToRootViewController(animated: false)
```
내비게이션컨트롤러를 이렇게 까지 가져올 필요가 있을까요!? 코멘트에 남긴 프로퍼티들을 활용해보세요.
여기서 popToRoot는 불필요해보입니다. 코멘트에 첨언할게용 :)

myssun0325 commented on CloudNotes/CloudNotes/Controllers/ListViewController.swift:
```diff
@@ -0,0 +1,162 @@
+    private func showContentViewController(indexPath: IndexPath?) {
+        let contentVC = ContentViewController()
+        if indexPath != nil {
+            contentVC.didTapMemoItem(with: memoList[indexPath!.row])
```
옵셔널을 체크하고 아니라면 강제언래핑을 하기보단 처음부터 옵셔널 바인딩으로 접근하는게 좋아보입니다.
contentVC에 didTapMemoItem으로 메모를 넘기는 순간 안에서 UI도 업데이트를 하는게 좋을까요? (아직 contentVC가 화면에 보이기도 전입니다)

iluxsm replied:
우선 parameter로 사용된 `indexPath`을 옵셔널로 선언한 이유부터 설명드리면, 값이 필요없는 부분에서도 해당 메서드가 호출될 수 있기 때문에 일부러 옵셔널 타입으로 선언했습니다. 그리고 argument가 `nil`이냐 아니냐에 따라 작업을 분류하기 위해 해당 조건문을 사용했습니다. 그리고 UI업데이트 부분에 관해서는 의도한 구현이라고 말씀드렸기에 생략하겠습니다!

myssun0325 replied:
아 제가 의도했던건 indexPath가 옵셔널인지 아닌지를 판단하고 안에서 그걸 강제언랩핑할꺼라면 아래와 같은건 어떤지 였습니다.
```swift
if let indexPath = indexPath {
  contentVC.didTapMemoItem(with: memoList[indexPath.row])
} else {
}
```

myssun0325 commented on CloudNotes/CloudNotes/Controllers/ContentViewController.swift:
```diff
@@ -0,0 +1,244 @@
+    }
+    
+    
+    private func modifyContents() -> Memo?  {
```
내부에서 하려는 행동이 contentView의 text를 가지고 Memo를 만들어 주기 위함이라면
실질적으로 필요한 text를 input으로 받고 output으로 Memo를 만들어주는 메서드로 개선해보는건 어떨까요? 그렇게 되면 메서드가 contentView에 의존하지 않게 되겠죠!?

myssun0325 commented on CloudNotes/CloudNotes/Controllers/ContentViewController.swift:
```diff
@@ -0,0 +1,244 @@
+    
+    private func modifyContents() -> Memo?  {
+        let startIndex: String.Index = contentView.text.startIndex
+        guard let endIndex: String.Index = contentView.text.firstIndex(of: "\n") else { return Memo() }
```
인덱스를 활용해서 문자열을 짤라오는 방법을 굉장히 잘 활용하신거 같네용 👍
근데 정말 else 구문을 만났을 때 빈메모라도 반환해야하는걸까요 아니면 반환타입으로 활용하신 옵셔널을 활용해야하는걸까요?
(+ 참고하기: swift에서 문자열을 짜르기 위한 여러방법들이 있습니다. 참고해보세용 split이라든지, components라든지. 각 메서드가 어떤 장단점이 있을까요?~ 이번단계에서 이걸 모두 정리할 필요가 없습니다. 나중에 문자열관련해서 공부하게 되면 꼭 챙겨가면 좋습니다. 실제 현업에서도 문자열을 굉장히 많이 다루게 될겁니다.!)

myssun0325 commented on CloudNotes/CloudNotes/Controllers/ListViewController.swift:
```diff
@@ -0,0 +1,162 @@
+    }
+}
+extension ListViewController {
+    //MARK: CREATE
```
CRUD에 대한 메서드들을 잘 작성해주었네요~👏👏👏
내부에서 save와 fetchMemo를 모든메서드에서 해주는거라면 어떻게 중복을 줄일 수 있을까요?

myssun0325 commented on CloudNotes/CloudNotes/Extensions/DateExtension.swift:
```diff
@@ -0,0 +1,17 @@
+import Foundation
+extension Date {
+    func convertToDate() -> String {
+        let dateFormatter = DateFormatter()
```
이렇게 매번 DateFormatter를 생성하는것보단
사용하는 곳에서만 특정포맷을 사용한다면 사용처에서 DateFormatter를 들고있는건 어떤가요?
보니깐 특정뷰컨에서 특정목적으로만 사용되고있군용

iluxsm replied:
저는 굳이 `Date` 의 메서드로 추가할 필요가 없다는 말로 이해했는데, 맞는지 잘 모르겠습니다! 하지만 그 말이 맞다면, 특정 뷰컨에서 특정 목적으로만 사용되고 있다는 상황에서는 말씀하신 방향이 더 맞는 것 같습니다 🤓

`+` 추가로 메이슨의 의견이 궁금한 것이 있는데, 만약 `converToDate()` 메서드가 여러 곳에서 사용되어야 한다면 (기획의 변경 등), Extension으로 `Date` 메서드로 추가한 것이 괜찮은 선택으로 보시나요? 그리고 `converToDate()`와 같이 나중을 생각해서 짠 코드와 현재 적재적소에 있어야 할 코드 중 어떤 것을 더 우선시 하는 것이 좋다고 생각하시나요?? 궁금합니다!

myssun0325 replied:
개인적인 의견으론 미래의 변경사항을 위해서 그때도 쓰일 코드를 예측해서 미리 짜두기보단 미래의 변경사항에 유연하게 수정/확장 할 수 있는 코드를 짜는것에 중점을 두는 편입니다.

언급해주신 메서드 자체만 어디에 두고 쓸지만 보자면 저라면 convertToDate()라는 메서드가 특정 뷰컨에서만 사용된다면 그 뷰컨의 메서드로 만들어두고 사용하거나(DateFormatter를 매번생성하도록은 X) 그 파일의 fileprivate으로 extension을 걸거나(또는 extension에 조건을 건다든지) 할거같네요.

나중에 다른 곳에서 필요한 경우에 그 경우에 맞게 다시 유지보수를 해야한다고 생각해요.

myssun0325 commented on CloudNotes/CloudNotes/Controllers/ContentViewController.swift:
```diff
@@ -0,0 +1,244 @@
+        ])
+    }
+    
+    private func updateUI(with memo: Memo) {
```
옵셔널로 인한 코드들을 조금 더 깔끔하게 할 수 있는 방법은 없을까요!? 매번 옵셔널을 확인하고있네요

myssun0325 commented on CloudNotes/CloudNotes/Controllers/ContentViewController.swift:
```diff
@@ -0,0 +1,244 @@
+    override func viewWillDisappear(_ animated: Bool) {
+        super.viewWillDisappear(animated)
+        
+        if let mainVC = self.splitViewController as? MainViewController,
```
mainVC -> mainViewController

myssun0325 commented on CloudNotes/CloudNotes/Controllers/ContentViewController.swift:
```diff
@@ -0,0 +1,244 @@
+        super.viewWillDisappear(animated)
+        
+        if let mainVC = self.splitViewController as? MainViewController,
+           let modifiedContents = modifyContents() {
```
[델리게이트를 활용하신 부분]
ListViewController에서 ContentViewController를 데이터를 넘길 때 ListViewController가 ContentViewController의 delegate가 되고, ContentViewController에서의 데이터 업데이트가 delegate로 데이터의 업데이트에 관한 정보를 넘겨주도록 구성해보는건 어떤가요? ContentViewController가 직접적으로 ListViewController에 대해서 알게되면 강한 커플링이 걸립니다.
ContentViewController는 델리게이트 객체가 프로토콜을 따른다는것만 알지 실질적으로 구체적인 타입(ListViewController)를 알지못합니당

<img width="642" alt="image" src="https://user-images.githubusercontent.com/26790375/110021796-89ff0c00-7d6e-11eb-83b7-21ce91a4aa01.png">

myssun0325 reviewed (commented):
수고하셨습니당. <kbd>글렌 이니 ❤️🧡💛💚💙남💜</kbd>
피드백과 함께 코멘트들을 확인해보세요.
(우리 같이해보기로했던 코드로 레이아웃해보기는 시간을 잡아봅시당 😄)

- NSAtrributedString을 잘 활용하셨습니다 👍
- 뷰를 다루는 코드들이 멋지게 발전하고 있는거 같습니다 :) 
- 삭제하기 버튼이 잘 기능합니다  :) 
- contentVC라는 약어보다는 ViewContorller의 장황하고 명확함을 선택해보는건 어떤가요?
- 내비게이션 스택에 대해서 잘 숙지하고 계신지 체크해보는건 어떤가요? ContentViewController로 push하기전에 pop을 하는거와 안하는건 어떤차이가 있을까요? (push만 해보고 동작을 비교해보세요)
- 델리게이트 패턴을 적극활용중이신데 순환참조가 일어날 수 있는 경우들에 대해 알아보고 어떻게 이를 해결하는지 알아보아용~
- ContentViewController의 UIAlertController의 Action들을 프로퍼티로 가지고 있어야할 필요가 있을까요? 불필요하게 프로퍼티가 많아지면 관리해야하는 것들이 많아진다는 것을 의미할 수도 있습니다. 개인적으론 이전에 다른 뷰컨에서 했던 방식처럼 UIAlertController를 세팅해주는 메서드로 빼는게 훨씬좋아보입니다.
- UIViewController에서 바로 접근할 수 있는 navigationController와 splitViewController 프로퍼티를 활용해보세요 ~
- 고민에 남겨주신거처럼 모델을 관리하는 객체를 따로 만들어서 CRUD를 담당하도록 해보는건 좋은아이디어 같습니다. 고고하시죠~! 🤠 MemoDataManager 기대해봅니다. (객체의 메서드를 작성할 때 정말 이 객체가 해야하는 일인지 다시 한 번 고민하면서 작성해보아요)
- 문자열 리터럴에 대해서는 충분히 디스코드로 얘기한거같아서 별도로 남기진않겠습니다 !? (핵심은 사용하는 곳에서 어떤 장점을 취해갈지 등)
- 고민에 남겨주신거처럼 여러분도 모르게 이제 기능이 커지거나 역할이 이상해지면 낌새를 알아차리는거 같군요 👍좋습니다
- 고민으로 남겨주셨던 부분에 대해서 모두 동의합니다.

myssun0325 commented:
@iluxsm 
> 문자열 리터럴을 사용하던 것을 StringCollection을 따로 만들어서 관리하도록 변경했습니다. 그런데, 기준을 잘 잡은 것인지 잘 모르겠네요. 개인적으로 문자열 리터럴이 변경될 수 있다고 생각하는 것들을 카테고리로 모아서 여러 객체로 만들어 놨는데, 현명한 선택인지 잘 모르겠습니다.
- 저는 현명한 선택이라고 생각합니다. 따로 만들어서 관리하되, 아까 구두로 설명한거처럼 사용하는 곳에 따라서 따로 `어떻게` 만들어서 관리할건지를 고민해보면 좋을거 같습니다. 추가적으로 카테고리로 모아서 여러 객체로 만들어둔 부분이 어떤점이 좋을거라고 생각해서 그렇게 하셨는지를 정리해둔다면 나중에 자신만의 기준을 세울 때 도움이 될거 같네용

> nil-coalescing 연산자 또는 Default로 빈 문자열("") 이 들어가는 것들을 String()으로 변경해줬습니다. 사실 이 것에 대해서 검색 해봤을 때, 내부적인 동작으로도 두 가지 방법이 동일한 것이다라는 것을 알게됐습니다. 하지만, 문자열 리터럴을 따로 관리해주는 입장에서 통일성(?)을 위해 해줬습니다. 이 부분에 대해서 메이슨의 생각이 궁금하네요.
- 빈 문자열 생성 자체에 대해서 리터럴을 쓸지 String()을 쓸지는 저도 글렌과 같은 의견입니다. 어떻게 생성해서 컨벤션을 통일하냐를 고민해봤다면 좀 더 나아가서 빈 문자열이 사용되는곳에서 매번 String()이나 ""로 빈문자열을 생성하는 것과 변수나 상수로 선언해두고 사용하는것에 대해서 고민해보면 좋을거 같습니다 ^^

Commit:
[#18] refactor: StringCollection 수정 및 네임 스페이스 활용

[Description]
- 빈 문자열 또는 줄바꿈 문자는 String의 extension에 추가했습니다.
- AlertController 라는 큰 열거형 안에 각각의 상황에 맞는 열거형을 중첩했습니다.
